File: src/hasoftware-library/src/hasoftware/api/messages/OutputDeviceRequest.java
package hasoftware.api.messages;

import hasoftware.api.FunctionCode;
import hasoftware.api.Message;
import hasoftware.api.classes.OutputDevice;
import hasoftware.cdef.CDEFAction;
import hasoftware.cdef.CDEFMessage;
import java.util.LinkedList;
import java.util.List;

public class OutputDeviceRequest extends Message {

    private int _action;
    private final List<Integer> _ids = new LinkedList<>();
    private final List<OutputDevice> _outputDevices = new LinkedList<>();

    public OutputDeviceRequest() {
        this(CDEFAction.None);
    }

    public OutputDeviceRequest(int action) {
        super(FunctionCode.OutputDevice, 0);
        _action = action;
    }

    public OutputDeviceRequest(CDEFMessage cdefMessage) {
        super(cdefMessage);
        _action = cdefMessage.getU8();
        int countIds = cdefMessage.getU8();
        for (int index = 0; index < countIds; index++) {
            _ids.add(cdefMessage.getU32());
        }
        int countOutputDevices = cdefMessage.getU8();
        for (int index = 0; index < countOutputDevices; index++) {
            _outputDevices.add(new OutputDevice(cdefMessage));
        }
    }

    public OutputDeviceResponse createResponse() {
        OutputDeviceResponse response = new OutputDeviceResponse(getTransactionNumber());
        response.setAction(_action);
        return response;
    }

    @Override
    public void encode(CDEFMessage cdefMessage) {
        super.encode(cdefMessage);
        cdefMessage.putU8(_action);
        cdefMessage.putU8(_ids.size());
        for (Integer id : _ids) {
            cdefMessage.putU32(id);
        }
        cdefMessage.putU8(_outputDevices.size());
        for (OutputDevice outputDevice : _outputDevices) {
            outputDevice.encode(cdefMessage);
        }
    }

    public int getAction() {
        return _action;
    }

    public void setAction(int action) {
        _action = action;
    }

    public List<Integer> getIds() {
        return _ids;
    }

    public List<OutputDevice> getOutputDevices() {
        return _outputDevices;
    }
}

File: src.net/hasoftware-library/Messages/OutputDeviceRequest.cs
using hasoftware.Api;
using hasoftware.Cdef;
using hasoftware.Classes;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace hasoftware.Messages
{
    public class OutputDeviceRequest : Message
    {
        public CdefAction Action { get; set; }
        public List<int> Ids { get; private set; }
        public List<OutputDevice> OutputDevices { get; private set; }

        public OutputDeviceRequest()
            : this(CdefAction.None)
        {
        }

        public OutputDeviceRequest(CdefAction action)
           : base(Api.FunctionCode.OutputDevice, 0)
        {
            Action = action;
            Ids = new List<int>();
            OutputDevices = new List<OutputDevice>();
        }

        public OutputDeviceRequest(CdefMessage cdefMessage)
            : base(cdefMessage)
        {
            Action = (CdefAction)cdefMessage.GetU8();
            Ids = new List<int>();
            int countIds = cdefMessage.GetU8();
            for (int index = 0; index < countIds; index++) {
                Ids.Add(cdefMessage.GetU32());
            }
            OutputDevices = new List<OutputDevice>();
            int countOutputDevices = cdefMessage.GetU8();
            for (int index = 0; index < countOutputDevices; index++) {
                OutputDevices.Add(new OutputDevice(cdefMessage));
            }
        }

        public OutputDeviceResponse CreateResponse() {
            OutputDeviceResponse response = new OutputDeviceResponse(TransactionNumber);
            response.Action = Action;
            return response;
        }

        public override void Encode(CdefMessage cdefMessage) {
            base.Encode(cdefMessage);
            cdefMessage.PutU8((int)Action);
            cdefMessage.PutU8(Ids.Count);
            foreach (var id in Ids) {
                cdefMessage.PutU32(id);
            }
            cdefMessage.PutU8(OutputDevices.Count);
            foreach (var outputDevice in OutputDevices) {
                outputDevice.Encode(cdefMessage);
            }
        }
    }
}
